dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,301 @@
+{
+    auto tokens = std::vector<MathToken>();
+
+    size_t index = 0;
+    bool error = false;
+    bool expectNumber = true;
+    while (input[0] != '\0' && !error)
+    {
+        // Skip spaces
+        auto nextChar = input[0];
+        if (nextChar != L' ')
+        {
+            if (expectNumber)
+            {
+                if (nextChar == '(')
+                {
+                    // Open parens are also acceptable, but don't change the next expected token type.
+                    tokens.push_back(MathToken(MathTokenType::Parenthesis, nextChar));
+                }
+                else
+                {
+                    double value = NAN;
+                    const auto charLength = GetNextNumber(input, numberParser, value);
+
+                    if (charLength > 0)
+                    {
+                        tokens.push_back(MathToken(MathTokenType::Numeric, value));
+                        input += charLength - 1; // advance the end of the token
+                        expectNumber = false; // next token should be an operator
+                    }
+                    else
+                    {
+                        // Error case -- next token is not a number
+                        error = true;
+                    }
+                }
+            }
+            else
+            {
+                if (c_numberBoxOperators.find(nextChar) != std::wstring::npos)
+                {
+                    tokens.push_back(MathToken(MathTokenType::Operator, nextChar));
+                    expectNumber = true; // next token should be a number
+                }
+                else if (nextChar == ')')
+                {
+                    // Closed parens are also acceptable, but don't change the next expected token type.
+                    tokens.push_back(MathToken(MathTokenType::Parenthesis, nextChar));
+                }
+                else
+                {
+                    // Error case -- could not evaluate part of the expression
+                    error = true;
+                }
+            }
+        }
+
+        input++;
+    }
+
+    if (error)
+    {
+        return {};
+    }
+    return tokens;
+}
+
+// Attempts to parse a number from the beginning of the given input string. Returns the character size of the matched string.
+size_t NumberBoxParser::GetNextNumber(std::wstring input, winrt::INumberParser numberParser, double& outValue)
+{
+    size_t charLength = 0;
+
+    // Attempt to parse anything before an operator or space as a number
+    std::wregex regex(L"^-?([^-+/*\\(\\)\\^\\s]+)");
+    std::wsmatch match;
+    if (std::regex_search(input.cbegin(), input.cend(), match, regex))
+    {
+        // Might be a number
+        const auto matchLength = match[0].length();
+        const auto parsedNum = numberParser.ParseDouble(input.substr(0, matchLength));
+
+        if (parsedNum)
```

Comment:
>parsedNum [](start = 12, length = 9)

What is this checking for? If 0 is returned by ParseDouble will we skip this block?

Reply:
No, ParseDouble returns an IReference<double>. This is checking for a null value, which means an error.

Reply:
Does an IReference<double> who's value is 0 return true to its boolean operator?

---
In reply to: [345447843](https://github.com/microsoft/microsoft-ui-xaml/pull/1602#discussion_r345447843) [](ancestors = 345447843)

Reply:
Yeah, I tried it. But I can add a test case so it's validated.